cmd/list.go

```diff
@@ -2,51 +2,110 @@ package cmd
 
 import (
 	"encoding/json"
+	"encoding/xml"
 	"fmt"
-	"gopkg.in/yaml.v2"
 	"io"
-	"os"
+
+	"github.com/ory/viper"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
 
 	"github.com/boson-project/faas"
 	"github.com/boson-project/faas/knative"
-	"github.com/spf13/cobra"
 )
 
-var formats = map[string]fmtFn{
-	"plain": fmtPlain,
-	"json":  fmtJSON,
-	"yaml":  fmtYAML,
+func init() {
+	root.AddCommand(listCmd)
+	listCmd.Flags().StringP("namespace", "n", "", "Override namespace in which to search for Functions.  Default is to use currently active underlying platform setting - $FAAS_NAMESPACE")
+	listCmd.Flags().StringP("output", "o", "plain", "optionally specify output format (plain,json,yaml)")
+
+	err := listCmd.RegisterFlagCompletionFunc("output", CompleteOutputFormatList)
+	if err != nil {
+		fmt.Println("Error while calling RegisterFlagCompletionFunc: ", err)
+	}
+}
+
+var listCmd = &cobra.Command{
+	Use:        "list",
+	Short:      "Lists deployed Functions",
+	Long:       `Lists deployed Functions`,
+	SuggestFor: []string{"ls", "lsit"},
+	PreRunE:    bindEnv("namespace", "output"),
+	RunE:       runList,
 }
 
-var validFormats []string
+func runList(cmd *cobra.Command, args []string) (err error) {
+	config := newListConfig()
 
-func completeFormats(cmd *cobra.Command, args []string, toComplete string) (formats []string, directive cobra.ShellCompDirective) {
-	formats = validFormats
-	directive = cobra.ShellCompDirectiveDefault
+	lister, err := knative.NewLister(config.Namespace)
+	if err != nil {
+		return
+	}
+	lister.Verbose = config.Verbose
+
+	client := faas.New(
+		faas.WithVerbose(verbose),
+		faas.WithLister(lister))
+
+	names, err := client.List()
+	if err != nil {
+		return
+	}
+
+	formatted, err := formatNames(names, config.Output)
+	if err != nil {
+		return
+	}
+
+	fmt.Println(formatted)
 	return
 }
 
-func init() {
-	root.AddCommand(listCmd)
+// TODO: placeholder. Create a fit-for-purpose Names plaintext formatter
+func fmtNamesPlain(i interface{}) ([]byte, error) {
```

Comment:
@lkingland  There was custom function before, wasn't it?
I see that signature has been unified.
TBH I liked 
```go
type fmtFn func(writer io.Writer, names []string) error
```
more, although it meant that the *.Marshal functions couldn't have been used directly.

Reply:
I like think you're right. When trying to chose between the two, I initially chose the one directly using the *.Marshal functions.  I've added back custom formatting functions, created a Format type, a serializer, and moved these funcitons to be methods of a custom type per output in [the most recent commit](https://github.com/boson-project/faas/pull/79/commits/3fc39aa7738cd0f4c86f13a57a4d9758125968d3).